sim/src/plugin_wrapper.rs

```diff
@@ -0,0 +1,157 @@
+use crate::filter_types::{CodeletType, Filter};
+use crate::sim_element::SimElement;
+use rpc_lib::rpc::Rpc;
+use std::collections::HashMap;
+use std::env;
+use std::fmt;
+use std::path::PathBuf;
+
+pub struct PluginWrapper {
+    // https://docs.rs/libloading/0.6.5/libloading/os/index.html
+    // TODO: Currently uses a platform-specific binding, which isn't very safe.
+    filter: Filter,
+    loaded_function: libloading::os::unix::Symbol<CodeletType>,
+    id: u32,
+    stored_rpc: Option<Rpc>,
+    neighbor: Option<u32>,
+}
+
+impl fmt::Display for PluginWrapper {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(width) = f.width() {
+            write!(
+                f,
+                "{:width$}",
+                &format!("PluginWrapper {{ id : {} }}", &self.id),
+                width = width
+            )
+        } else {
+            write!(f, "PluginWrapper {{ id : {} }}", &self.id)
+        }
+    }
+}
+
+impl SimElement for PluginWrapper {
+    fn tick(&mut self, _tick: u64) -> Vec<(Rpc, Option<u32>)> {
+        if self.stored_rpc.is_some() {
+            let ret = self.execute(self.stored_rpc.as_ref().unwrap());
+            self.stored_rpc = None;
+            if ret.is_none() {
+                vec![]
+            } else {
+                vec![(ret.unwrap(), self.neighbor)]
+            }
+        } else {
+            vec![]
+        }
+    }
+    fn recv(&mut self, rpc: Rpc, _tick: u64) {
+        assert!(self.stored_rpc.is_none(), "Overwriting previous RPC");
+        self.stored_rpc = Some(rpc);
+    }
+    fn add_connection(&mut self, neighbor: u32) {
+        self.neighbor = Some(neighbor);
+    }
+}
+
+fn load_lib(plugin_str: &str) -> libloading::Library {
```

Comment:
This is the biggest change. I wrote a custom loading function that loads libraries differently depending on whether you are working with Linux type systems or MacOS. Github workflows should test this, took me a bit to get this right.

Reply:
This makes sense, but can we add a comment that libloading::os::unix also works for mac?  Otherwise it can be confusing to read if you don't look it up.

Reply:
MacOS is a Unix distribution, although a fairly distant branch. Added a comment